src/OrchardCore/OrchardCore.DisplayManagement/Razor/RazorPage.cs

```diff
@@ -269,6 +272,50 @@ public new bool IsSectionDefined(string name)
             return zone != null && zone.Items.Count > 0;
         }
 
+        public new void DefineSection(string name, Microsoft.AspNetCore.Mvc.Razor.RenderAsyncDelegate section)
+        {
+            var zone = ThemeLayout.Zones[name];
+            dynamic sectionShape = New.AspSection().GetAwaiter().GetResult();
+            sectionShape.RenderAsyncDelegate = SectionDelegate(name, section, sectionShape);
+
+            if (zone is Zones.ZoneOnDemand zoneOnDemand)
+            {
+                zoneOnDemand.AddAsync(sectionShape).GetAwaiter().GetResult();
+            }
+            else
+            {
+                zone.Add(sectionShape);
+            }
+        }
+
+        private void BeginWriteSectionZone()
+        {
+            if (_writer == null)
+            {
+                _writer = new StringWriter();
+            }
+        }
+
+        private void EndWriteSectionZone()
+        {
+            _writer.GetStringBuilder().Clear();
+        }
+
+        private RenderAsyncDelegate SectionDelegate(string name, RenderAsyncDelegate oldDelegate, dynamic shape)
+        {
+            RenderAsyncDelegate newDelegate = async () =>
+            {
+                BeginWriteSectionZone();
+                PushWriter(_writer);
+                await oldDelegate();
+                PopWriter();
+                shape.HtmlContent = _writer.ToString();
+                EndWriteSectionZone();
+                shape.RenderAsyncDelegate = null;
+            };
+            return newDelegate;
+        }
+
```

Comment:
Was not able to change the files directly, just able to do a PR on your branch, so just copy / paste my suggested changes. Note: Here without the name parameter that need to be also removed on the call

        private RenderAsyncDelegate SectionDelegate(RenderAsyncDelegate section, dynamic shape)
        {
            return async () =>
            {
                BeginWriteSection();
                PushWriter(_writer);
                await section();
                PopWriter();
                shape.HtmlContent = _writer.ToString();
                EndWriteSection();
                shape.RenderAsyncDelegate = null;
            };
        }